### PR TITLE
fix posixct conv back and forth

### LINKED ISSUE #578: `$rolling()` count with temporal window ignores decimal seconds
I have a set of high-resolution time series data on which I need to perform rolling operations with time-based windows. Here is some sample data for replication purposes:

```r
library(polars) # package version '0.11.0'

## sample data
x = as.POSIXct(
  c(
    "2020-01-01 13:45:48.343", "2020-01-01 13:45:48.815"
    , "2020-01-01 13:45:49.289", "2020-01-01 13:45:49.974"
    , "2020-01-01 13:45:51.190", "2020-01-01 13:45:51.631"
  )
  , tz = "UTC"
)

## create {polars} `DataFrame`
df = pl$DataFrame(
  "ts" = x
)

## set sorted
df = df$with_columns(
  pl$col("ts")$set_sorted()
)
```

Now, when performing a simple rolling count over a 1-second moving window, the outcome looks as follows:

```r
## perform rolling count
df$with_columns(
  "n" = pl$count("ts")$
    rolling(
      index_column = "ts"
      , period = "1s"
      , closed = "both"
    )
)

# shape: (6, 2)
# ┌─────────────────────────┬─────┐
# │ ts                      ┆ n   │
# │ ---                     ┆ --- │
# │ datetime[ms, UTC]       ┆ u32 │
# ╞═════════════════════════╪═════╡
# │ 2020-01-01 13:45:48 UTC ┆ 2   │
# │ 2020-01-01 13:45:48 UTC ┆ 2   │
# │ 2020-01-01 13:45:49 UTC ┆ 4   │
# │ 2020-01-01 13:45:49 UTC ┆ 4   │
# │ 2020-01-01 13:45:51 UTC ┆ 2   │
# │ 2020-01-01 13:45:51 UTC ┆ 2   │
# └─────────────────────────┴─────┘
```

Given the sample data, I would expect the resulting count to start at `1L` and then increase up to `3L` at the 3rd position as more values fall in the 1-second lookback period. However, it seems like decimal seconds are disregarded entirely and all timestamps within the full second are considered instead.

I also tested this in Python using the exact same sample data, which gives the expected result:

```py
import polars as pl # package version '0.19.19'

df = (
  pl.DataFrame(
    {
      "ts": [
        "2020-01-01 13:45:48.343", "2020-01-01 13:45:48.815", 
        "2020-01-01 13:45:49.289", "2020-01-01 13:45:49.974", 
        "2020-01-01 13:45:51.190", "2020-01-01 13:45:51.631"
      ]
    }
  )
  # convert to datetime
  .with_columns(
    pl.col("ts").str.to_datetime(time_unit = "ms", time_zone = "UTC")
  )
  # set sorted
  .set_sorted("ts")
)

# perform rolling count
df.with_columns(
  pl.count("ts")
    .rolling(index_column = "ts", period = "1s", closed = "both")
    .alias("n")
)

# shape: (6, 2)
# ┌─────────────────────────────┬─────┐
# │ ts                          ┆ n   │
# │ ---                         ┆ --- │
# │ datetime[ms, UTC]           ┆ u32 │
# ╞═════════════════════════════╪═════╡
# │ 2020-01-01 13:45:48.343 UTC ┆ 1   │
# │ 2020-01-01 13:45:48.815 UTC ┆ 2   │
# │ 2020-01-01 13:45:49.289 UTC ┆ 3   │
# │ 2020-01-01 13:45:49.974 UTC ┆ 2   │
# │ 2020-01-01 13:45:51.190 UTC ┆ 1   │
# │ 2020-01-01 13:45:51.631 UTC ┆ 2   │
# └─────────────────────────────┴─────┘
```

What's strange is that in R, decimal seconds are not printed despite the indicated temporal resolution `ms` &ndash; something that cannot be resolved through overriding the print behavior for decimal seconds via `options("digits.secs" = 3L)`. Otherwise, I do not see an apparent difference between the two versions.

### PR DESCRIPTION
**Pull request by sorhawell**
resolves #578

### CODE REVIEW

File: tests/testthat/test-expr_datetime.R [hunk 1]
```diff
@@ -18,6 +18,18 @@ test_that("pl$lit posix", {
...
+
+  x = as.POSIXct(
+    c(
+      "2020-01-01 13:45:48.343", "2020-01-01 13:45:48.815"
+      , "2020-01-01 13:45:49.289", "2020-01-01 13:45:49.974"
+      , "2020-01-01 13:45:51.190", "2020-01-01 13:45:51.631"
+    ), tz = "UTC"
+  )
+  expect_identical(pl$lit(x)$to_r(), x)  #preserve millisecond precision
+  expect_failure(expect_identical(pl$lit(x)$to_r(), x+0.001)) #control, detect 1ms offset
```
Comment (eitsupi) [inline tests/testthat/test-expr_datetime.R:31]: I feel this test should not be here. This is not a test for expr, but for type conversion.

However, we did not know where such a test is currently put together.

**Review by eitsupi — APPROVED** [review]
Thanks for this update.

@sorhawell I tweak the tests. Does this make sense?

File: tests/testthat/test-datatype.R [hunk 2]
```diff
@@ -32,3 +32,37 @@ test_that("plStruct", {
...
+
+  expect_identical(
+    pl$lit(as.POSIXct("2022-01-01", tz = "HST"))$to_r(),
+    as.POSIXct("2022-01-01", tz = "HST")
+  )
+
+  expect_identical(
+    pl$lit(as.POSIXct("2022-01-01", tz = "GMT"))$to_r(),
+    as.POSIXct("2022-01-01", tz = "GMT")
+  )
```
Comment (etiennebacher) [inline tests/testthat/test-datatype.R:56]: this test is the same as the second one

Comment (eitsupi) [reply tests/testthat/test-datatype.R:56]: Testing regarding data type mapping definitely needs to be improved.
I think we can merge this into a follow-up task for now.

Comment (etiennebacher) [reply tests/testthat/test-datatype.R:56]: As you want but adding a `# TODO` or removing the identical test takes the same time ;)

Comment (eitsupi) [reply tests/testthat/test-datatype.R:56]: I created an issue #592

**Pull request merged**